crates/bitwarden-crypto/src/opaque_ke/opaque.rs

```diff
@@ -0,0 +1,374 @@
+    ) -> Result<types::ServerRegistrationFinishResult, Error> {
+        let registration = ServerRegistration::finish(RegistrationUpload::<Self>::deserialize(
+            registration_upload,
+        )?);
+        Ok(types::ServerRegistrationFinishResult {
+            server_registration: registration.serialize().to_vec(),
+        })
+    }
+
+    fn start_client_login(&self, password: &str) -> Result<types::ClientLoginStartResult, Error> {
+        let result = ClientLogin::<Self>::start(&mut rand::thread_rng(), password.as_bytes())?;
+        Ok(types::ClientLoginStartResult {
+            credential_request: result.message.serialize().to_vec().into(),
+            state: result.state.serialize().to_vec().into(),
+        })
+    }
+    fn start_server_login(
+        &self,
+        server_setup: &[u8],
+        server_registration: &[u8],
+        credential_request: &[u8],
+        username: &str,
+    ) -> Result<types::ServerLoginStartResult, Error> {
+        let result = ServerLogin::start(
+            &mut rand::thread_rng(),
+            &ServerSetup::<Self>::deserialize(server_setup)?,
+            Some(ServerRegistration::<Self>::deserialize(
+                server_registration,
+            )?),
+            CredentialRequest::deserialize(credential_request)?,
+            username.as_bytes(),
+            ServerLoginStartParameters::default(),
+        )?;
+        Ok(types::ServerLoginStartResult {
+            credential_response: result.message.serialize().to_vec(),
+            state: result.state.serialize().to_vec(),
+        })
+    }
+    fn finish_client_login(
+        &self,
+        state: &[u8],
+        credential_response: &[u8],
+        password: &str,
+    ) -> Result<types::ClientLoginFinishResult, Error> {
+        let client_login = ClientLogin::<Self>::deserialize(state)?;
+        let result = client_login.finish(
+            password.as_bytes(),
+            CredentialResponse::deserialize(credential_response)?,
+            ClientLoginFinishParameters::new(None, Identifiers::default(), Some(&self.get_ksf()?)),
+        )?;
+        Ok(types::ClientLoginFinishResult {
+            credential_finalization: result.message.serialize().to_vec().into(),
+            session_key: result.session_key.to_vec().into(),
+            export_key: result.export_key.to_vec().into(),
+            server_public_key: result.server_s_pk.serialize().to_vec().into(),
+        })
+    }
+    fn finish_server_login(
+        &self,
+        state: &[u8],
+        credential_finalization: &[u8],
+    ) -> Result<types::ServerLoginFinishResult, Error> {
+        let server_login = ServerLogin::<Self>::deserialize(state)?;
+        let result = server_login.finish(CredentialFinalization::deserialize(
+            credential_finalization,
+        )?)?;
+        Ok(types::ServerLoginFinishResult {
+            session_key: result.session_key.to_vec(),
+        })
+    }
+}
+
+#[cfg(test)]
+mod test {
```

Comment:
Should we add any unhappy path tests? My thinking is the error that we return on a 
wrong password should be well documented and tested so that we can handle that on the client side and show a nice error.